lib/rex/post/meterpreter/packet.rb

```diff
@@ -693,17 +819,31 @@ def to_r
   # passing it on to the default functionality that can parse
   # the TLV values.
   #
-  def from_r(bytes)
-    xor_key = bytes[0,4].unpack('N')[0]
-    super(xor_bytes(xor_key, bytes[4, bytes.length]))
+  def from_r(aes_key=nil)
+    xor_key = self.raw[XOR_KEY_OFF, XOR_KEY_SIZE]
+    data = xor_key + xor_bytes(xor_key, self.raw[SESSION_GUID_OFF, self.raw.length - SESSION_GUID_OFF])
+    self.session_guid = data[SESSION_GUID_OFF, SESSION_GUID_SIZE]
+    encrypted_flag = data[ENCRYPTED_FLAG_OFF, ENCRYPTED_FLAG_SIZE] == "\x01"
+    packet_length = data[PACKET_LENGTH_OFF, PACKET_LENGTH_SIZE]
+    packet_type = data[PACKET_TYPE_OFF, PACKET_TYPE_SIZE]
+    self.type = packet_type.unpack('N')[0]
+    if encrypted_flag
```

Comment:
There are scenarios, as per the PR description, where plain text TLVs are needed. There may be more.

The whole "trust the protocol" thing falls back into the "you have to trust something" category that I've already discussed. If you can't trust that byte, you can't trust any bytes. you can't trust the session GUID, or the length, or anything. So yes, I think trusting that byte is ok as far as an active 'toggle this byte' MITM is concerned.

I get the concern for having an unencrypted TLV during an encrypted session though, as far as passive replays go. It may be enough for MSF to say "if I have a key, I expect encryption" and go from there.

This will have to be considered more before settling on it though. Good question.